// netschoolapi/dist/methods/logIn/methods/schoolInfo.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
function default_1(client, school) {
    return __awaiter(this, void 0, void 0, function* () {
        const schools = yield client
            .get("addresses/schools")
            .then((res) => res.json());
        if (!schools || !schools.length)
            throw new Error("Сетевой не вернул список школ");
        if (typeof school === "string")
            school = school.toLowerCase();
        const result = schools.find(({ id, name }) => typeof school !== "string" ? id == school : name.toLowerCase() == school);
        if (!result)
            throw new Error("Не удалось найти школу");
        return {
            cid: result.countryId,
            sid: result.stateId,
            pid: result.municipalityDistrictId,
            cn: result.cityId,
            sft: 2,
            scid: result.id,
        };
    });
}
exports.default = default_1;
