// netschoolapi/dist/classes/Assignment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Assignment {
    constructor(assignment) {
        var _a, _b, _c, _d, _e, _f;
        this.id = assignment.id;
        this.dot = (_b = (_a = assignment.mark) === null || _a === void 0 ? void 0 : _a.dutyMark) !== null && _b !== void 0 ? _b : false;
        this.text = assignment.assignmentName;
        this.mark = (_d = (_c = assignment.mark) === null || _c === void 0 ? void 0 : _c.mark) !== null && _d !== void 0 ? _d : null;
        this.typeId = assignment.typeId;
        this.comment = (_f = (_e = assignment.markComment) === null || _e === void 0 ? void 0 : _e.name) !== null && _f !== void 0 ? _f : null;
        this._date = assignment.dueDate;
        this.lessonId = assignment.classMeetingId;
    }
    get date() {
        return new Date(this._date);
    }
    toJSON() {
        return {
            id: this.id,
            dot: this.dot,
            text: this.text,
            mark: this.mark,
            typeId: this.typeId,
            lessonId: this.lessonId,
        };
    }
}
exports.default = Assignment;
