// netschoolapi/dist/classes/Context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Context {
    constructor(credentials) {
        this.user = credentials.user;
        this.year = credentials.year;
        this.server = credentials.server;
        this.school = credentials.school;
        this.subjects = credentials.subjects;
    }
    /** Проверяет является ли число частью года */
    checkDate(date) {
        const { start, end } = this.year;
        return +start <= +date && +date <= +end;
    }
    /** Существует ли класс */
    classExists(id) {
        return !!this.user.classes.find((c) => c.id == id);
    }
    /** ID первого класса */
    defaultClass() {
        var _a, _b;
        return (_b = (_a = this.user.classes[0]) === null || _a === void 0 ? void 0 : _a.id) !== null && _b !== void 0 ? _b : -1;
    }
    /** Существует ли ученик */
    studentExists(id) {
        return !!this.user.students.find((s) => s.id == id);
    }
    /** ID первого ученика */
    defaultStudent() {
        var _a, _b;
        return (_b = (_a = this.user.students[0]) === null || _a === void 0 ? void 0 : _a.id) !== null && _b !== void 0 ? _b : -1;
    }
    /** Существует ли предмет */
    subjectExists(id) {
        return !!this.subjects.find((s) => s.id == id);
    }
}
exports.default = Context;
